### ISSUE TITLE
Weight Clipper: set constraints on parameters or layers

### ISSUE THREAD

**Issue by L-M-Sherlock** [burn-rs/burn#662]
### Feature description

I need to keep the model's weights within a certain range during training, so that they don't go too high or too low.

### Feature motivation

Some weights in my model are very sensitive. If they go out of a specific range during training, the result will be NaN.

### (Optional) Suggest a Solution

PyTorch allows for clamping weights during training. You can find more on this topic in this post: https://discuss.pytorch.org/t/set-constraints-on-parameters-or-layers/23620/2

### Supplement

I am in the process of rewriting one of my PyTorch-based projects using Burn. Here's the code I'm using in Python:

```python
class WeightClipper:
    def __init__(self, frequency: int=1):
        self.frequency = frequency

    def __call__(self, module):
        if hasattr(module, 'w'):
            w = module.w.data
            w[4] = w[4].clamp(1, 10)
            w[5] = w[5].clamp(0.1, 5)
            w[6] = w[6].clamp(0.1, 5)
            w[7] = w[7].clamp(0, 0.5)
            w[8] = w[8].clamp(0, 3)
            w[9] = w[9].clamp(0.1, 0.8)
            w[10] = w[10].clamp(0.01, 2.5)
            w[11] = w[11].clamp(0.5, 5)
            w[12] = w[12].clamp(0.01, 0.2)
            w[13] = w[13].clamp(0.01, 0.9)
            w[14] = w[14].clamp(0.01, 2)
            w[15] = w[15].clamp(0, 1)
            w[16] = w[16].clamp(1, 10)
            module.w.data = w
```
Source: https://github.com/open-spaced-repetition/fsrs-optimizer/blob/95694b787bb71ac9883db1201af09e334ee5ee0b/src/fsrs_optimizer/fsrs_optimizer.py#L100-L120

```python
            for i, batch in enumerate(self.next_train_data_loader):
                self.model.train()
                self.optimizer.zero_grad()
                sequences, delta_ts, labels, seq_lens = batch
                real_batch_size = seq_lens.shape[0]
                outputs, _ = self.model(sequences)
                stabilities = outputs[seq_lens-1, torch.arange(real_batch_size), 0]
                retentions = power_forgetting_curve(delta_ts, stabilities)
                loss = self.loss_fn(retentions, labels).sum()
                loss.backward()
                for param in self.model.parameters():
                    param.grad[:4] = torch.zeros(4)
                self.optimizer.step()
                self.scheduler.step()
                self.model.apply(self.clipper)
                pbar.update(real_batch_size)
```
Source: https://github.com/open-spaced-repetition/fsrs-optimizer/blob/95694b787bb71ac9883db1201af09e334ee5ee0b/src/fsrs_optimizer/fsrs_optimizer.py#L231-L246

**Comment by antimora** [comment]
Relevant issues:
1. Clamp feature PR (recently added; it's on the main branch): https://github.com/burn-rs/burn/pull/550
2. Making weights public (still pending): https://github.com/burn-rs/burn/issues/594

The clamp feature automatically determines whether it's an in-place update (if there are no references) or a light clone.

One thing I'm uncertain about is the clamp operation on trainable tensors. It will track changes, but since we are clipping after training, should this be disabled? @nathanielsimard will know more about this requirement.

**Comment by nathanielsimard** [comment]
I'm not sure you will be able to modify the current module parameters during training, since those can be used by multiple threads. What you could do is wrap `Param<Tensor>` inside a `Mutex` or a `RwLock` to safelly modify their value. 

I don't think `Mutex` and `RwLock` implement `Module` and `ADModule` trait, so we should add them here first: https://github.com/burn-rs/burn/blob/main/burn-core/src/module/param/primitive.rs

**Comment by L-M-Sherlock** [comment]
I'm not sure how will you implement this feature. In my torch project, the Weight Clipper will be applied after optim.step(). I find that model is mutable in the training epoch:

https://github.com/burn-rs/burn/blob/6b5ba770847ae0b4a2dda6bc584415c6c8008270/burn-train/src/learner/epoch.rs#L110-L122

**Comment by nathanielsimard** [comment]
It's true that you could call a mutable method on your module right after the gradients are applied. You wouldn't need any lock to do that, but as of today, you would need to implement your own training loop. Maybe we could introduce a way to register mutable callbacks.

**Comment by dae** [comment]
@nathanielsimard i'd be willing to sponsor a few hours on this and other issues related to porting an open-source PyTorch project over to burn - if you're interested, please contact me at d-burn at ankitects.com

**Comment by L-M-Sherlock** [comment]
I have implemented Weight Clipper in my project with PR https://github.com/open-spaced-repetition/fsrs-rs/pull/35. Thanks for all contributors here.
